misk-core/src/main/kotlin/misk/queuing/TimedBlockingQueue.kt

```diff
@@ -0,0 +1,171 @@
+package misk.queuing
+
+import java.time.Duration
+import java.util.concurrent.ArrayBlockingQueue
+import java.util.concurrent.BlockingQueue
+import java.util.concurrent.TimeUnit
+
+/**
+ * This BlockingQueue implementation allows measuring how long dequeued items have been
+ * in the queue. It requires a delayHandler which consumes the latency every time an item is
+ * removed from the queue.
+ * Note that the the actual removal of an item from the queue and the delayHandler invocation
+ * are not atomic, hence it may not be appropriate for applications which require
+ * a higher level of accuracy.
+ * @param <T> The type of the elements held in this queue
+ */
+class TimedBlockingQueue<T>(
+  maxQueueSize: Int,
+  private val delayHandler: (Duration) -> Unit
+) : BlockingQueue<T> {
+  private val queue: BlockingQueue<TimedQueueItem<T>> =
+      ArrayBlockingQueue<TimedQueueItem<T>>(maxQueueSize)
+
+  override fun poll(timeout: Long, unit: TimeUnit): T? {
+    val item = queue.poll(timeout, unit)
+    return if (item != null) handleRemoved(item) else null
+  }
+
+  override fun poll(): T? {
+    val item = queue.poll()
+    return if (item != null) handleRemoved(item) else null
+  }
+
+  override fun take(): T {
+    return handleRemoved(queue.take())
+  }
+
+  override fun remove(): T {
+    return handleRemoved(queue.remove())
+  }
+
+  override fun contains(element: T): Boolean {
+    return queue.contains(wrap(element))
+  }
+
+  override fun addAll(elements: Collection<T>): Boolean {
+    return queue.addAll(wrapCollection(elements))
+  }
+
+  override fun clear() {
+    invokeDelayHandlerOnAll(queue)
```

Comment:
This isn’t atomic, so two concurrent calls could trigger multiple calls into the delay hander. Instead repeatedly call poll ?

Reply:
The same problem applies to `remove(element: T)`, `removeAll()` and `retainAll()`. Will think about this and see if we can solve it in a more generic way.

Reply:
Here are my thoughts:

We can use `poll()` for `clear()` as suggested. 

For `removeAll(collection: Collection<T>)`, `retainAll(collection: Collection<T>)`, we can follow a similar approach by calling `remove(element)` for each item in the collection. `remove(element)` in turn will invoke the delayHandler and the underlying `remove(element)`. This will work but since the delayHandler and the `remove(element)` calls are not protected by a lock, they can become out of sync. 

However, since the primary use case is for request queues, I think it is fine to leave as is. For example, the jetty QueuedThreadPool does not use the above operations. I have added comments about the limitations so that users are aware.

Reply:
I like the idea of keeping this internal-only and not spending too much energy implementing general-purpose features.

Reply:
Agreed. Will move this to misk-web and make internal only.